tests/integration/clientv3/naming/resolver_test.go

```diff
@@ -64,55 +63,93 @@ func TestEtcdGrpcResolver(t *testing.T) {
+	// Create connection with provided lb policy
+	conn, err := grpc.Dial("etcd:///foo", grpc.WithInsecure(), grpc.WithResolvers(b),
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingPolicy":"%s"}`, lbPolicy)))
 	if err != nil {
 		t.Fatal("failed to connect to foo", err)
 	}
 	defer conn.Close()
 
+	// Send an initial request that should go to e1
 	c := testpb.NewTestServiceClient(conn)
 	resp, err := c.UnaryCall(context.TODO(), &testpb.SimpleRequest{}, grpc.WaitForReady(true))
 	if err != nil {
 		t.Fatal("failed to invoke rpc to foo (e1)", err)
 	}
-	if resp.GetPayload() == nil || !bytes.Equal(resp.GetPayload().GetBody(), s1PayloadBody) {
+	if resp.GetPayload() == nil || !bytes.Equal(resp.GetPayload().GetBody(), payloadBody) {
 		t.Fatalf("unexpected response from foo (e1): %s", resp.GetPayload().GetBody())
 	}
 
-	em.DeleteEndpoint(context.TODO(), "foo/e1")
-	em.AddEndpoint(context.TODO(), "foo/e2", e2)
+	// Send more requests
+	lastResponse := []byte{'1'}
+	totalRequests := 100
+	for i := 1; i < totalRequests; i++ {
+		resp, err := c.UnaryCall(context.TODO(), &testpb.SimpleRequest{}, grpc.WaitForReady(true))
+		if err != nil {
+			t.Fatal("failed to invoke rpc to foo", err)
+		}
 
-	// We use a loop with deadline of 30s to avoid test getting flake
-	// as it's asynchronous for gRPC Client to update underlying connections.
-	maxRetries := 300
-	retryPeriod := 100 * time.Millisecond
-	retries := 0
-	for {
-		time.Sleep(retryPeriod)
-		retries++
+		t.Logf("Response: %v", string(resp.GetPayload().GetBody()))
+
+		if resp.GetPayload() == nil {
+			t.Fatalf("unexpected response from foo: %s", resp.GetPayload().GetBody())
+		}
+		lastResponse = resp.GetPayload().GetBody()
+	}
 
-		resp, err = c.UnaryCall(context.TODO(), &testpb.SimpleRequest{})
+	// If the load balancing policy is pick first then return payload should equal number of requests
+	if lbPolicy == "pick_first" {
+		t.Logf("Last response: %v", string(lastResponse))
+		if string(lastResponse) != "100" {
+			t.Fatalf("unexpected total responses from foo: %s", string(lastResponse))
+		}
+	}
+
+	// If the load balancing policy is round robin we should see roughly half total requests served by each server
+	if lbPolicy == "round_robin" {
+		t.Logf("Last response: %v", string(lastResponse))
+		responses, err := strconv.Atoi(string(lastResponse))
 		if err != nil {
-			if retries < maxRetries {
-				continue
-			}
-			t.Fatal("failed to invoke rpc to foo (e2)", err)
+			t.Fatalf("couldn't convert to int: %s", string(lastResponse))
 		}
-		if resp.GetPayload() == nil || !bytes.Equal(resp.GetPayload().GetBody(), s2PayloadBody) {
-			if retries < maxRetries {
-				continue
-			}
-			t.Fatalf("unexpected response from foo (e2): %s", resp.GetPayload().GetBody())
+
+		// Allow 10% tolerance a round robin is not perfect and we don't want the test to flake
+		threshold := float64(totalRequests) * 0.5 * 0.9
+		if responses < int(threshold) {
```

Comment:
stretchr has a nice function for deltas:

> assert.InDelta(t, math.Pi, 22/7.0, 0.01)

in case you want to make use of the upper boundary too.

Reply:
Great suggestion, added upper bound via stretchr, thanks! 🙏🏻